### PR TITLE
Use IMetaStoreClient in CachingHiveMetastore (issue 3764)

### PR DESCRIPTION
**Pull request by mbeitchman**
Change the CachingHiveMetastore to use the IMetaStoreClient interface
instead of directly calling the ThriftHiveMetastore client. This allows
for the CachingHiveMetastore implementation to use other implementations
of the IMetastoreClient interface without having to change any code
in the CachingHiveMetastore implementation.

The HiveMetastoreClient is now a mostly complete implementation of
IMetaStoreClient even though the CachingHiveMetastore implementation
only uses 14 of the 120 methods in the interface.

CLA signed via Amazon corporate CLA. I am listed on the schedule A.

### CODE REVIEW

File: presto-docs/src/main/sphinx/overview/concepts.rst [hunk 1]
```diff
@@ -151,11 +151,11 @@ Statement
 ^^^^^^^^^
 
 Presto executes ANSI-compatible SQL statements.  When Presto
-documentation refers to a statement we are refering to statements as
+documentation refers to a statement we are referring to statements as
```
Comment (electrum) [inline presto-docs/src/main/sphinx/overview/concepts.rst]: Thanks for finding and fixing these typos. Can you move this to a separate commit, since it's unrelated to the Hive change?

Comment (mbeitchman) [inline presto-docs/src/main/sphinx/overview/concepts.rst]: sure

Comment (mbeitchman) [inline presto-docs/src/main/sphinx/overview/concepts.rst]: done

**Comment by electrum** [comment]
Can you explain more about the motivation for this change? Do you have another metastore client implementation that needs retry and caching logic from `CachingHiveMetastore`? If you don't need that behavior, we already have the `HiveMetastore` interface (which at some point should probably be refactored not to use the Hive Thrift objects).

Otherwise, I think a better approach is to make `HiveMetastoreClient` an interface by pulling up the methods that are used by `CachingHiveMetastore`. Basically, it would have the same methods as `MockHiveMetastoreClient`:
```java
public interface HiveMetastoreClient
        implements Closeable
{
    @Override
    void close(); // no IOException

    List<String> get_all_databases()
            throws TException;

    ...
}
```
(We could later refactor the method names to use standard Java naming instead of snake case.)

The advantage of this approach is having a cleaner interface and moves our code away from relying on Hive interfaces, making it easier to implement alternative metastore clients. As you noted, `IMetastoreClient` has over a hundred methods and is not easy to implement. Implementing `HiveMetastoreClient` by delegating to `IMetastoreClient` would be a lot easier than the other way around.

We've actually wanted to have other metastore implementations. For example, we could have something that reads flat files on S3, allowing you to just point Presto at an S3 bucket and start querying.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java [hunk 2]
```diff
@@ -314,18 +314,20 @@ public void flushCache()
     private List<String> loadAllDatabases()
             throws Exception
     {
+        Optional<IMetaStoreClient> metastoreClient = Optional.of(clientProvider.createMetastoreClient());
         try {
             return retry()
                     .stopOnIllegalExceptions()
                     .run("getAllDatabases", stats.getGetAllDatabases().wrap(() -> {
-                        try (HiveMetastoreClient client = clientProvider.createMetastoreClient()) {
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java]: Acquiring the client inside the retry handler is important: it allows connecting to a different metastore on failure.

Comment (mbeitchman) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java]: The caching and retry logic is generally beneficial which is why I took this approach. I also took this approach because it allows someone to not only use a different metastore implementation easily with Presto (by just updating the factory) but also with Hive.

Comment (mbeitchman) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java]: I get the point about acquiring the client in the retry. I'll update the pr.

**Comment by mbeitchman** [comment]
I updated this pr to initialize the client in the retry handler but I am happy to go with your approach of making HiveMetastoreClient an interface. I'll send a new pr with that change shortly. Does that sound good?
